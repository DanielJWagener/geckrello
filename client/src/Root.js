// Simply creates a Redux store and passes it in to a provider tag.
// This logic was originally in index.js, but we've movied it here to
// make our testing process easier.
import React from "react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware, compose } from "redux";
import reduxThunk from "redux-thunk";
import reducers from "./reducers";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  reducers,
  {},
  composeEnhancers(applyMiddleware(reduxThunk))
);

export default ({ children }) => {
  return <Provider store={store}>{children}</Provider>;
};
