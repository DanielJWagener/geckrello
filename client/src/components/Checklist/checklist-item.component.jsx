import React from "react";
import { connect } from "react-redux";
import {
  checkOrUncheckChecklistItem as checkOrUncheck,
  deleteChecklistItem as deleteItem
} from "../../redux/cards/cards.actions";

// Conditionally render a checked or unchecked checklist item
export const ChecklistItem = ({
  checkOrUncheck,
  deleteItem,
  cardId,
  checklistItemId,
  checked,
  itemLabel
}) => (
  <div className="item">
    <div
      className={`item__checkbox ${
        checked ? "item__checkbox--checked" : "item__checkbox--unchecked"
      }`}
      onClick={() => checkOrUncheck(cardId, checklistItemId, checked)}
    >
      {checked ? <div className="checkmark">&#10004;</div> : <></>}
    </div>
    <div className="item__title">{itemLabel}</div>
    <div
      className="item__delete"
      onClick={() => deleteItem(cardId, checklistItemId)}
    >
      &times;
    </div>
  </div>
);

export default connect(null, {
  checkOrUncheck,
  deleteItem
})(ChecklistItem);
