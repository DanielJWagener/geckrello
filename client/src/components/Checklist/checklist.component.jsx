import React from "react";
import { connect } from "react-redux";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck } from "@fortawesome/free-solid-svg-icons";
import { addChecklistItem } from "../../actions";

import ChecklistItem from "./checklist-item.component";

import "./checklist.styles.scss";
import "./widget.styles.scss";

export class Checklist extends React.Component {
  state = {
    mode: "prompt",
    newItemInput: "",
    deleteCount: 0,
  };

  toggleMode = () => {
    if (this.state.mode === "prompt") {
      this.setState({ mode: "input" });
    } else if (this.state.mode === "input") {
      this.setState({ mode: "prompt" });
    }
  };

  onFormSubmit = e => {
    e.preventDefault();

    if (!this.state.newItemInput) {
      return false;
    }

    //Redux call to add new item to checklist array of card (unchecked by default)
    this.props.addChecklistItem(
      this.props.cardId,
      this.state.newItemInput,
      false
    );

    //Clear input from state
    this.setState({ newItemInput: "" });
  };

  handleChange = e => {
    this.setState({ newItemInput: e.target.value });
  };

  checklistItemsArray = () =>
    this.props.card.checklist.map(item => (
      <ChecklistItem
        key={item._id}
        itemLabel={item.label}
        checklistItemId={item._id}
        cardId={this.props.cardId}
      />
    ));

  render() {
    const addChecklistItemForm = (
      <form onSubmit={this.onFormSubmit} className="modal-form">
        <div className="modal-form__group">
          <input
            type="text"
            className="modal-form modal-form__input modal-form__input--validation"
            value={this.state.newItemInput}
            onChange={this.handleChange}
            style={{ borderBottomColor: this.state.borderColor }}
            placeholder="Add checklist item"
            autoFocus
          />
        </div>
        <div className="modal-form__group">
          <input
            type="submit"
            className="modal-form__submit"
            value="Add Item"
          />
          <button className="modal-form__cancel" onClick={this.toggleMode}>
            Cancel
          </button>
        </div>
      </form>
    );

    return (
      <div className="checklist">
        <div className="widget-heading">
          <FontAwesomeIcon className="widget-heading__icon" icon={faCheck} />
          <h3 className="widget-heading__text">Checklist</h3>
          <button onClick={this.toggleMode} className="widget-heading__button">
            Add Item
          </button>
        </div>

        <div className="checklist__items">{this.checklistItemsArray()}</div>
        {this.state.mode === "input" ? addChecklistItemForm : <></>}
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  const { cardId } = ownProps;
  const card = state.cards.filter(card => card._id === cardId)[0];
  return { card, cards: state.cards };
};

export default connect(mapStateToProps, { addChecklistItem })(Checklist);
