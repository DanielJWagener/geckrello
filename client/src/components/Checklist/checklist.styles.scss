.checklist {
  width: 100%;
  margin-bottom: 4rem;

  &__items {
    .item {
      display: flex;
      align-items: center;

      &__checkbox {
        //display: inline-block;
        height: 1.8rem;
        width: 1.8rem;
        border-radius: 2px;
        border: 2px solid #888;
        margin-right: 2rem;
        cursor: pointer;

        &--unchecked {
          background-color: white;
        }

        &--checked {
          position: relative;
          background-color: blue;
          color: white;
          font-size: 1.5rem;

          .checkmark {
            position: absolute;
            top: 50%;
            left: 50%;
            transform: translate(-50%, -50%);
          }
        }
      }

      &__title {
        //display: inline-block;
        font-size: 1.8rem;
      }

      &__delete {
        margin-left: 2rem;
        font-size: 1.5rem;
        cursor: pointer;

        &:hover {
          color: red;
        }
      }
    }
  }
}
