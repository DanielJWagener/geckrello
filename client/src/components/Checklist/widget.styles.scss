.widget-heading {
  margin-bottom: 0.5rem;
  display: grid;
  grid-template-columns:
    [heading-start icon-start] 7% [icon-end text-start] min-content
    [text-end button-start] minmax(min-content, auto) [button-end heading-end];

  &__icon {
    justify-self: start;
    align-self: center;
    font-size: 1.7rem;
    grid-column: heading-start / icon-end;
  }

  &__text {
    font-size: 2rem;
    margin-right: 1rem;
    grid-column: text-start / text-end;
  }

  &__button {
    grid-column: button-start / heading-end;
    justify-self: start;
    align-self: center;
    background-color: #888;
    color: white;
    font-size: 1.2rem;
    padding: 3px 1rem 3px 1rem;

    border-radius: 0.5rem;
    border: #888;
    cursor: pointer;

    &:hover {
      background-color: #444;
    }
  }
}
