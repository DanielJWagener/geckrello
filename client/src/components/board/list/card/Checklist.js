import React from "react";
import { connect } from "react-redux";
import { addChecklistItem } from "../../../../actions";

import ChecklistItem from "./ChecklistItem";

class Checklist extends React.Component {
  state = {
    mode: "prompt",
    newItemInput: "",
    deleteCount: 0
  };

  toggleMode = () => {
    if (this.state.mode === "prompt") {
      this.setState({ mode: "input" });
    } else if (this.state.mode === "input") {
      this.setState({ mode: "prompt" });
    }
  };

  onFormSubmit = e => {
    e.preventDefault();

    if (!this.state.newItemInput) {
      return false;
    }

    // Generate a unique key and listItemId in the event of duplicate list item titles
    const generateChecklistItemId = inputId => {
      if (
        !this.props.card.checklist.find(
          listItem => listItem.checklistItemId === inputId
        )
      ) {
        return inputId;
      } else {
        inputId = `${inputId}-duplicate`;
        return generateChecklistItemId(inputId);
      }
    };
    let checklistItemId = generateChecklistItemId(this.state.newItemInput);

    //Redux call to add new item to checklist array of card (unchecked by default)
    this.props.addChecklistItem(
      this.props.cardId,
      this.state.newItemInput,
      checklistItemId,
      false
    );

    //Clear input from state
    this.setState({ newItemInput: "" });
  };

  handleChange = e => {
    this.setState({ newItemInput: e.target.value });
  };

  signalDelete = () => {
    let newDeleteCount = this.state.deleteCount + 1;
    this.setState({ deleteCount: newDeleteCount });
  };

  checklistItemsArray = () =>
    this.props.card.checklist.map(item => (
      <ChecklistItem
        key={item.checklistItemId}
        checklistItemTitle={item.checklistItemTitle}
        checklistItemId={item.checklistItemId}
        cardId={this.props.cardId}
        signalDelete={this.signalDelete}
      />
    ));

  render() {
    if (this.state.mode === "prompt") {
      return (
        <div className="checklist">
          <h3 className="modal__widget-heading">Checklist</h3>
          <button onClick={this.toggleMode} className="modal__button--inline">
            Add Item
          </button>
          <div className="checklist__items">{this.checklistItemsArray()}</div>
        </div>
      );
    } else {
      return (
        <div className="checklist">
          <h3 className="modal__widget-heading">Checklist</h3>
          <div className="checklist__items">{this.checklistItemsArray()}</div>
          <form onSubmit={this.onFormSubmit} className="form">
            <div className="form__group">
              <input
                type="text"
                className="form form__input form__input--validation"
                value={this.state.newItemInput}
                onChange={this.handleChange}
                style={{ borderBottomColor: this.state.borderColor }}
                placeholder="Add checklist item"
                autoFocus
              />
            </div>
            <div className="form__group">
              <input type="submit" className="form__submit" value="Add Item" />
              <button className="form__cancel" onClick={this.toggleMode}>
                Cancel
              </button>
            </div>
          </form>
        </div>
      );
    }
  }
}

const mapStateToProps = (state, ownProps) => {
  const { cardId } = ownProps;
  const card = state.cards.filter(card => card.cardId === cardId)[0];
  return { card };
};

export default connect(
  mapStateToProps,
  { addChecklistItem }
)(Checklist);
