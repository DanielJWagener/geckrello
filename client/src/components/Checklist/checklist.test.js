import { shallow } from "enzyme";
import React from "react";
import Checklist from "./checklist.component";
import Root from "../../Root";

it("renders Checklist component", () => {
  expect(
    shallow(
      <Root>
        <Checklist />
      </Root>
    )
  ).toMatchSnapshot();
});
