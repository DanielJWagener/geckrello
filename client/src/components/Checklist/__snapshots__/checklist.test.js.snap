// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Checklist component 1`] = `ShallowWrapper {}`;
